File: y2018/python/day1.py
freq_changes = [int(i) for i in open("input/day1.txt").readlines()]

print(f"final freq: {sum(freq_changes)}")

dupe_found = False
freq = 0
seen = {freq}
while True:
    for fc in freq_changes:
        freq += fc
        if freq in seen:
            print(f"first dupe freq: {freq}")
            dupe_found = True
            break
        seen.add(freq)

    if dupe_found:
        break

File: y2018/rust/src/bin/day1.rs
use std::collections::HashSet;
use std::fs;

fn main() {
    let chances: Vec<_> = fs::read_to_string("../input/day1.txt")
        .unwrap()
        .split_whitespace()
        .map(|i| i.parse::<i32>().unwrap())
        .collect();

    let mut curr_freq = 0;
    let mut seen_freqs = HashSet::from([curr_freq]);
    let mut seen_twice = i32::MIN;
    'simulation:
    loop {
        for f in &chances {
            curr_freq += f;
            if seen_freqs.contains(&curr_freq) && seen_twice == i32::MIN {
                seen_twice = curr_freq;
                break 'simulation;
            }
            seen_freqs.insert(curr_freq);
        }
    }

    println!("final frequency: {}", chances.iter().sum::<i32>());
    println!("first frequency seen twice: {seen_twice}");
}
